src/core/internal/hash.d

```diff
@@ -30,6 +30,86 @@ private enum floatCoalesceNaNs = true;
         assert(hashOf(double.nan) == hashOf(-double.nan)); // Same hash for different NaN.
 }
 
+private template hasCallableToHash(T)
+{
+    static if (is(typeof(((T* x) => (*x).toHash())(null)) HashType))
+        enum hasCallableToHash = __traits(isIntegral, HashType) && is(HashType : size_t);
```

Comment:
>If we allow return types that are not exactly size_t, but convert to size_t, we may as well allow other kinds of subtyping (e.g. struct with alias this).

If you're comfortable with supporting `alias this` then `is(typeof(...) : size_t)` is what I'd write. The problem with checking if `cast(size_t) val.toHash()` compiles is that the cast throws away information for some types, and so the result is not what the writer of the `toHash` intended (for use in whatever context it made sense) but something potentially much less reasonable.

Reply:
> If you're comfortable with supporting `alias this` then `is(typeof(...) : size_t)` is what I'd write.

Yeah, I would like to give you the same suggestion, but unfortunately `is(typeof(<some expr>))` doesn't imply that `<some expr>` will compile, even though the compiler may give you a legit looking type.

https://forum.dlang.org/post/k6o04g$259s$1@digitalmars.com

> The problem with checking if `cast(size_t) val.toHash()` compiles is that the cast throws away information for some types, and so the result is not what the writer of the toHash intended 

Can you provide an example? My intention was to precisely specify the minimal possible requirements for the type `T`, which is equivalent to the way you call `toHash` in the code.

Reply:
> If you're comfortable with supporting alias this [..]

To clarify, my intention is minimal restrictions for the implementation of `T.toHash`. However that was provoked by the use of `: size_t`, instead of `== size_t`. I'm not sure I see an advantage of requiring an exact signature.

Reply:
>Can you provide an example?

For instance if `toHash` returned a floating point number.

Reply:
But is that a reasonable and useful design? To be honest I don't even immediately see high value in a struct with `alias this`, though I think allowing any arithmetic type could be interesting - e.g. `CheckedInt`.

Reply:
>But is that a reasonable and useful design? 

No I don't think so at all! If `toHash` is a float then it must have been intended for some unrelated purpose. Per "duck typing", it doesn't quack like a duck so I don't want to try to use it like a duck.

>I think we can't get away from the requirement on toHash's return value to behave like an integer.

I completely agree.

Reply:
Removed the `__traits(isIntegral, HashType)` requirement.